linkerd/app/core/src/errors.rs

```diff
@@ -19,3 +19,16 @@ pub fn root_cause<'e>(
     }
     error
 }
+
+/// Determines if the provided error was caused by an `E` typed error.
+pub fn caused_by<'e, E: std::error::Error + 'static>(
+    mut error: &'e (dyn std::error::Error + 'static),
+) -> Option<&'e E> {
+    while let Some(src) = error.source() {
+        if let Some(e) = src.downcast_ref::<E>() {
+            return Some(e);
+        }
+        error = src;
+    }
+    None
+}
```

Comment:
very minor nit: i might consider putting this in `linkerd-error`, so it could later be used in other crates. also, we already have https://github.com/linkerd/linkerd2-proxy/blob/a8e4a2e3f713d7417b51cd8ffd408d481823e49c/linkerd/error/src/lib.rs#L18-L20 which i think could be replaced with this function in a few places.

(not a big deal, i just feel like we've written a bunch of variants of this function in different places...)

Reply:
I'll followup with a cleanup